test/test_optim.py

```diff
@@ -1892,5 +1893,29 @@ def test_bn_update_eval_momentum(self):
         self.assertEqual(dnn.bn.momentum, 0.3)
 
 
+# Device-generic tests. Instantiated below and not run directly.
+class TestOptimDeviceType(TestCase):
+
+    # see https://github.com/pytorch/pytorch/issues/42364
+    def test_adamax_step(self, device):
+        shape = (2, 3, 4, 5)
+
+        class ModuleWithNonContigWeights(torch.nn.Module):
+            def __init__(self):
+                super(ModuleWithNonContigWeights, self).__init__()
+                self.weight = torch.nn.Parameter(torch.randn(shape, device=device)
+                                                 .to(memory_format=torch.channels_last))
+
+            def forward(self, input: torch.Tensor) -> torch.Tensor:
+                return input * self.weight
+
+        m = ModuleWithNonContigWeights()
+        input = torch.randn(shape, device=device, requires_grad=True)
+        m(input).sum().backward()
```

Comment:
that's a lot of faith in shape propagation working as expected. Can you assert that gradients are actually discontiguous?

Reply:
I can, it's not totally clear how the test is supposed to be fixed if that assert fails though.

Reply:
I guess I can pre-create the discontiguous inputs, which should reduce the chance of that failing.

Reply:
or even better, I can just create the grad tensor without even invoking autograd.